# Coding Temple's Data Analytics Course
---
## Python II: Numpy & CSV's

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>


6) <b>Numpy</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Python List Comparison <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) In-Class Exercise #6 <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Creating an NDArray <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.array() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.zeros() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.ones() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.arange() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Making Lists into NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Performing Calculations on NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Summation <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Difference <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Multiplication <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Division <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Standard Distribution <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Mean <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Median <br> 
 &nbsp;&nbsp;&nbsp;&nbsp; f) Numpy Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Multi-dimensional Arrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Indexing NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Checking NDArray Type <br>
 &nbsp;&nbsp;&nbsp;&nbsp; j) Altering NDArray Type <br>
 &nbsp;&nbsp;&nbsp;&nbsp; k) Checking the Shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp; l) Altering the Shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp; m) In-Class Exercise #7 <br>
 &nbsp;&nbsp;&nbsp;&nbsp; n) Complex Indexing & Assigning <br>
 &nbsp;&nbsp;&nbsp;&nbsp; o) Elementwise Multplication <br>
 &nbsp;&nbsp;&nbsp;&nbsp; p) np.where() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; q) Random Sampling <br>

7) <b>Working With CSV's</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Imports <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Reading a CSV <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Loading a CSV's Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Checking Number of Records <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
# Let's take a look at how we can creaete a function to add two to a variable
def add_two(x):
    return x + 2

#Then we would test it:
print(add_two(4))

# This is great when I am looking to reuse a function. But what if I only needed it for a single use app?
# This is where lambda can come into play!

a_var = 4

# lambda >input variable<: whatever we want to do
print((lambda x: x+2)(a_var))

6
6


#### Saving to a Variable

In [3]:
# We can save lambda functions to a variable to make them more reusaable:

f_test = lambda x: x+2

f_test(4)

6

In [4]:
f_test(78)

80

#### Multiple Inputs

In [5]:
# What if we had 3 numbers and we wanted to be able to take the product of all those numbers?

#Exactly the same as a normal function:
def product(x,y,z):
    return x*y*z

print((lambda x,y,z: z*y*x)(3,5,8))

120


#### Passing a Lambda into a Function

In [6]:
def multiply(function, num):
    """
    function expects to see a fucntion. (lambda fucntion)
    num expest to see a single integer input
    """
    return function(num)
print(multiply(lambda x: x*x, 4))

#not only can I do this with lambda, but I can use ordinary functions as well
def squared_nums(num):
    return num **2

print(multiply(squared_nums,4))

16
16


#### Returning a Lambda from a Function

In [8]:
def multiply_test(num):
    return num *4

# Take this function and return it inside another function
def return_func():
    test =4
    def multiply(num):
        return num *2
    return multiply(test)
print(return_func())

# I can do the same thing useing lambda

def return_lambs(b,c):
    return lambda x,a: x+a+b+c

r_lamb = return_lambs(4,5)
print(r_lamb)
print(r_lamb(5,9))

8
<function return_lambs.<locals>.<lambda> at 0x000002686147BD80>
23


#### If Statements within Lambdas

In [10]:
# lambda functions are set up to be able to teak in conditional statements
# lambda x: TRUE if (conditional) else False
# lambda x: do something if (conditional is TRUE) else, do something else

f_conditional = lambda num: num*2 if num > 10 else num +2
print(f_conditional(8))
print(f_conditional(20))


10
40


In [12]:
lst = [1,2,3,4,5,6,7,8,9,10,11,12,13,14]
# Run the lambda on the list:
[f_conditional(num) for num in lst]


[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 22, 24, 26, 28]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
#Write out the function
anon_func = lambda x: x**3

# Invoke the function
anon_func(5) 

125

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [16]:
numbers = [4,11,20,3,15,20,11,3,2,6,5,1,44,1]

#Normally, in order to use a function on this, we would need to iterate through:
def squared_normal(val):
    return val **2
lst = []
for num in numbers:
    lst.append(squared_normal(num))

print(lst)

[16, 121, 400, 9, 225, 400, 121, 9, 4, 36, 25, 1, 1936, 1]


In [2]:
''' 
map(function, iterable)
At iterable is something we can loop over, such as: a list, dict, tuple, set or a string!
Normally, you would use map with a predefined fuction, but we can also use it with a lambda statement!
'''

def squared_map(num, num1):
    if num < 10 and num1 < 10:
        return num **2, num1 **2
    
    else:
        return num, num1
    
numbers = [4,11,20,3,15,20,11,3,2,6,5,1,44,1]
more_numbers = [4,10,3,2,6,10,15,3,4,7,6,0,42,1]

#In order to use the squared_map function without a map statement, we would need to iterate through both lists at once.
# to do so effectively, we will utilize the zip() fucntion without a map statement, which brings together two iterable objects and
# Returns the values at each index position, up till the end of the shortest length of iterables.

# for val1, val2 in zip(numbers, more_numbers):
#     print(squared_map(val1,val2))

list(map(squared_map, numbers, more_numbers))



[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (9, 9),
 (4, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 42),
 (1, 1)]

#### Using Lambda's with Map

In [122]:
"""`
Instead of having to build out a fucntion we plan on using 1 time
We can use a lambda inside the map fuction in itself instead. this allows us 
to do everything we need to do in a single line of code


map(lambda x:do something, iterable)
"""
# remember to always use the list function with the map function
#Run lambda over two lists, based on conditional
list(map(lambda x,y: (x**2,y**2) if x <10 and y<10 else (x,y) , numbers, more_numbers))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (9, 9),
 (4, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 42),
 (1, 1)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [22]:
list(map(lambda x: x*2 -1, numbers))

[7, 21, 39, 5, 29, 39, 21, 5, 3, 11, 9, 1, 87, 1]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [23]:
# Syntax looks like:
# list(filter(function, iterable(s)))
# A lot like a mask

names = 'Bob', 'Andy', 'Max', 'evan', 'Angela'

def a_names(name):
    # in this function we will check to see if each letter of each name begins with 'a'
    if name[0].lower()=='a':
        return True
    else:
        return False

#Our function here returns a boolean as value, just like any other conditional fucntion
#Filters take a conditional statement to filter out what is not desired

list(filter(a_names, names))  


['Andy', 'Angela']

#### Using Lambda's with Filter()

In [24]:
# WE can also use lambda's
# this is great if you are trying to filter data with a sinle-use function instead of having to define one

# When we use a lmbda, or comprehension, we will typically write out the value first, then the conditional
# The best way to think about this is like a reversed for loop

list(filter(lambda name: True if name[0].lower()== 'a' else False, names))


['Andy', 'Angela']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [124]:
import statistics
statistics.mean(numbers)

list(filter(lambda num: True if num >= statistics.mean(numbers) else False, numbers)) 

[11, 20, 15, 20, 11, 44]

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [32]:
# start off with defining the fucntion
def add_nums_breakable(num):
    #Without a base case, we will end up infitelely looping through our program
    if num>=20000:
        return num + add_nums_breakable(num -1)
    else:
        return num

In [35]:
def add_nums(num):
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
        return num + add_nums(num -1)
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [37]:
# 5! = 5*4*3*2*1

def factorial(num):
    if num <= 1:
        print('factorial(1) =(1)')
        return num
    else:
        print(f'factorial({num}) = {num} * factorial({num - 1})')
        return num * factorial(num - 1)
print(factorial(5))

factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
factorial(1) =(1)
120


#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [39]:

def subtract(num1, num2):
    if num2 ==0:
        return num1
    else:
        print(num1, num2)
        return subtract(num1 -1, num2 -1)
    
subtract(25,5)

25 5
24 4
23 3
22 2
21 1


20

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [45]:
[i for i in range(10)]

def my_range(stop, start=0, step=1):
    while start < stop:
        yield start
        start += step
print([i for i in my_range(10)])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [48]:
def squared_nums(num):
    while num > 0:
        yield num **2
        num -=1
    
[i for i in squared_nums(5)]

[25, 16, 9, 4, 1]

## Numpy <br>

<p>NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.</p>
<ul>
    <li>Shape = Rows & Columns</li>
    <li>Matrix = Entire Array</li>
    <li>Vector = Variables to be applied (same vector as the one used in physics)</li>
    <li>Array = Similar to lists</li>
</ul>

#### Python List Comparison

Lists are flexible, dynamic python objects that do their job quite well. But they do not support some mathematical operations in an intuitive way. Consider the summation of two lists, $l_1$ and $l_2$

In [50]:
l1 = [1,2,3]
l2 = [4,5,6]

print(f'Sum of the lists: {l1+l2}')

# Apply a subtraction element-wise 

Sum of the lists: [1, 2, 3, 4, 5, 6]


<p>If we wanted to sum lists elementwise, we could write our own function that does the job entirely within the framework of python</p>

#### In-Class Exercise #6 - Write a function that sums the indexes of two lists <br>
<p>Ex: [2, 3, 4] + [1, 5, 2] = [3, 8, 6]</p>

In [129]:
list1 = [5,5,10]
list2 = [2,4,1]

# Output: [7,9,11]
def sum_list(l1,l2):
    final = []
    i = 0 
    while i < len(l1):
        final.append(l1[i] + l2[i])
        i += +1
    return final
print(sum_list(list1,list2))

[7, 9, 11]


We would have to write a similar function for all the possible operands that we could consider for list arithmatic. This is time consuming and inefficient. Moreover, once the lists in question become nested, mimicing the behavior of true matrices, the problem gets worse. Complicated indexing is necessary, just to allow for the most basic matrix operations common throughout science and engineering. Imagine writing a matrix multiplication function using python syntax in a general way, such that it returns a matrix-matrix or matrix-vector product:

\begin{align}
(n \times x) \times (x \times m) \rightarrow (n \times m)
\end{align}

\begin{align}
\begin{bmatrix}
c_{0,0} & ... & c_{0,n} \\
\vdots & \ddots & \vdots \\
c_{m,0} & ... & c_{m,n}
\end{bmatrix}
=
\begin{bmatrix}
a_{0,0} & ... & a_{0,x} \\
\vdots & \ddots & \vdots \\
a_{n,0} & ... & a_{n,x}
\end{bmatrix}
\begin{bmatrix}
b_{0,0} & ... & b_{0,m} \\
\vdots & \ddots & \vdots \\
b_{x,0} & ... & b_{x,m}
\end{bmatrix}
\end{align}

Another Example:

<img src="https://www.mathsisfun.com/algebra/images/matrix-multiply-a.svg">

---

The "Dot Product" is where we multiply matching members, then sum up:

(1, 2, 3) • (7, 9, 11) = 1×7 + 2×9 + 3×11
    = 58

We match the 1st members (1 and 7), multiply them, likewise for the 2nd members (2 and 9) and the 3rd members (3 and 11), and finally sum them up.

<hr>

So, let's instantiate a matrix $\mathcal{M}$ and a vector $\vec{v}$ and write a function that does the multiplication ourselves.

In [132]:
def matrix_multiplication(A,B):
    zero_list = [[0 for i in range(len(B[0]))] for i in range(len(A))]
    print(zero_list)
    inner_dimensions = len(A[0])
    print(inner_dimensions)
    n_dimension = len(zero_list)
    print(n_dimension)
    m_dimension = len(zero_list[0])
    print(m_dimension)


#now we can loop though our matrix
    for i in range(n_dimension):
        for j in range(m_dimension):
            element = 0
            for x in range(inner_dimensions):
                element += A[i][x] * B[x][j]
            zero_list[i][j] = element
    return zero_list
print(matrix_multiplication([[0,1,0],[0,2,0],[0,3,0]],[[1],[2],[3]]))

[[0], [0], [0]]
3
3
1
[[2], [4], [6]]


#### Importing

In [61]:
import numpy as np


#### Creating an NDArray <br>
NumPy is based around a class called the $\textit{NDArray}$, which is a flexible vector / matrix class that implements the intuitive matrix and vector arithmatic lacking in basic Python. Let's start by creating some NDArrays:

##### - np.array()

In [62]:
arr_1 = np.array([1,2,3])
print(arr_1)
print(type(arr_1))

[1 2 3]
<class 'numpy.ndarray'>


##### - np.zeros()

In [64]:
# Create a shape variable, it will reference the 2-D shape I want my matrix to have
shape = (3,3)
# We create the matrix
M= np.zeros(shape, int)
M

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]])

##### - np.ones()

In [65]:
arr2 = np.ones(shape, int)
print(arr2)

[[1 1 1]
 [1 1 1]
 [1 1 1]]


##### - np.arange()

In [66]:
np.arange(1,12)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

##### - Making Lists into NDArrays

In [68]:
l3 = [4,5,6]

# we jsut use the .array() function to turn this list into a array instead:
arr_list = np.array(l3)
arr_list

array([4, 5, 6])

#### Performing Calculations on NDArrays

##### - Summation

In [69]:
arr5 = np.array([5,10,15,20,25])
arr10 = np.array([10,20,30,40,50])

arr5 + arr10


array([15, 30, 45, 60, 75])

##### - Difference

In [70]:
arr10 - arr5

array([ 5, 10, 15, 20, 25])

##### - Multiplication

In [71]:
arr5 * arr10

array([  50,  200,  450,  800, 1250])

##### - Division

In [72]:
arr10 / arr5

array([2., 2., 2., 2., 2.])

##### - Standard Deviation

In [73]:
print(np.std(arr10))

14.142135623730951


##### - Mean

In [75]:
int(np.mean(arr5))

15

##### - Median

In [76]:
np.median(arr5)

15.0

#### Numpy Subsetting

In [77]:
mask = arr10 >20
arr10[mask]

array([30, 40, 50])

#### Multi-dimensional Arrays <br>
NumPy seamlessly supports multidimensional arrays and matrices of arbitrary dimension without nesting NDArrays. NDArrays themselves are flexible and extensible and may be defined with such dimensions, with a rich API of common functions to facilitate their use. Let's start by building a two dimensional 3x3 matrix by conversion from a nested group of core python lists $M = [l_0, l_1, l_2]$:

In [133]:
a_list = [0,1,2]
b_list = [3,4,5]
c_list = [6,7,8]

# Step 1: Convert the list into a matrix that we can actually work with
M = [a_list, b_list, c_list]
print(f'This is a nested list structure: {M}')

# Step 2: we can cast it directly intal matrix row
# When I do so, the dimensions of that matrix are inferred by Numpy for us
M = np.array(M)
print(f'Matrix Structured: {M}')


This is a nested list structure: [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
Matrix Structured: [[0 1 2]
 [3 4 5]
 [6 7 8]]


#### Indexing NDArrays <br>
<p> Similar to lists within lists; however, the syntax looks more like C programming language.... It is [1, 2] to access the second row, third element.</p>

In [80]:
#  Because of the 2-D structure, single level indexing is going to return the value in the entire row:
print(f'Get the first array out of the matrix: {M[0]}')

# Multi-level indexing
print(f'Get the middel value of the array: {M[1][1]}')

Get the first array out of the matrix: [0 1 2]
Get the middel value of the array: 4


#### Assigning Values in NDArrays

In [88]:
# Multi - level way:
# This will separates out the statement in to two indicies
M[1][1] = 9
print(M)

# Giving a list of indicies:
M[0,1]= 8
print(M)

[[0 7 2]
 [3 9 5]
 [6 7 8]]
[[0 8 2]
 [3 9 5]
 [6 7 8]]


<p>Notice above how we ended up with a 1 in the target element's place. This is a data type issue. The .dtype() method is supported by all NDArrays, as well as the .astype() method for casting between data types:</p>

#### Checking NDArray Type

In [89]:
M.dtype

dtype('int32')

#### Altering NDArray Type

In [91]:
# Let's take a look at the type of the data before we make any changes:
t= M
print(f'Type before changes: {t.dtype}')

#Change the data to be float instead?
# We wil utlize the .astype() function

changed = t.astype(np.float64)
print(changed)
print(f'Type after the change: {changed.dtype}')

Type before changes: int32
[[0. 8. 2.]
 [3. 9. 5.]
 [6. 7. 8.]]
Type after the change: float64


#### Checking the Shape <br>
The behavior and properties of an NDArray are often sensitively dependent on the $\textit{shape}$ of the NDArray itself. The shape of an array can be found by calling the .shape method, which will return a tuple containing the array's dimensions:

In [92]:
# In a 2-D data structure, the shapes is the displayed as: (rows, columns) or (rows x columns)

print(changed.shape)

# 1 -D Vector
print(arr5.shape)

(3, 3)
(5,)


#### Altering the Shape <br>
<p>As long as the number of elements remains fixed, we can reshape NDArrays at will:</p>

In [94]:
print(changed.shape)
print(f'Now we will reshape the matrix into Vector: {changed.reshape(9,1).astype(int)}')

(3, 3)
Now we will reshape the matrix into Vector: [[0]
 [8]
 [2]
 [3]
 [9]
 [5]
 [6]
 [7]
 [8]]


#### In-Class Exercise #7 - Create a matrix of range 0 up to 16 and reshape it into a 4x4 matrice

In [3]:
import numpy as np

#### Complex Indexing & Assinging

In [4]:
# reset the matrixx

M = np.zeros((6,6)).astype(int)
print('Resetting the Matrix... Please wait')
print('Before assing values: \n {M}')

M[:0] =1
print(f'end=n After assigning 1 to the firs to column:\m{M}')

M[2:5, ] =5
print(f'\n After assitng 5:\n{M}')

#Reset back to zero
M = M*0
print(f'Reset was confirmed\n{M}')

#set the 2nd and 3rd columns to qual to the value of 2
M[:, 1:3]=2
print(f'Assigning 2 to the second and thrid columns:\n{M}')

Resetting the Matrix... Please wait
Before assing values: 
 {M}
end=n After assigning 1 to the firs to column:\m[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]

 After assitng 5:
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [5 5 5 5 5 5]
 [5 5 5 5 5 5]
 [5 5 5 5 5 5]
 [0 0 0 0 0 0]]
Reset was confirmed
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]
Assigning 2 to the second and thrid columns:
[[0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]]


#### Elementwise Multiplication

<p>As long as the shapes of NDArrays are 'compatible', they can be multiplied elementwise, broadcasted, used in inner products, and much much more. 'Compatible' in this context can mean compatible in the linear algebraic sense, i.e. for inner products and other matrix multiplication, or simply sharing a dimension in such a manner that broadcasting 'makes sense'. Here are some examples of this:</p>

In [8]:
M *= 0 

matrix = M
#Create tjhe vector of range 6:
vector = np.arange(6)
print(matrix)
print('\n')
print(vector)


matrix += vector
print(matrix)

matrix*= vector
print(matrix)

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]


[0 1 2 3 4 5]
[[0 1 2 3 4 5]
 [0 1 2 3 4 5]
 [0 1 2 3 4 5]
 [0 1 2 3 4 5]
 [0 1 2 3 4 5]
 [0 1 2 3 4 5]]
[[ 0  1  4  9 16 25]
 [ 0  1  4  9 16 25]
 [ 0  1  4  9 16 25]
 [ 0  1  4  9 16 25]
 [ 0  1  4  9 16 25]
 [ 0  1  4  9 16 25]]


#### np.where() <br>
<p>If statement within NDArrays that allows you to run conditionals on the entire array</p>

In [9]:
print(f'{np.where(M>2)}')

mask = np.where(M==1)
M[mask] = 2
print(M)

(array([0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5,
       5, 5], dtype=int64), array([2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3,
       4, 5], dtype=int64))
[[ 0  2  4  9 16 25]
 [ 0  2  4  9 16 25]
 [ 0  2  4  9 16 25]
 [ 0  2  4  9 16 25]
 [ 0  2  4  9 16 25]
 [ 0  2  4  9 16 25]]


#### Random Sampling <br>
<p>NumPy provides machinery to work with random numbers - something often needed in a broad spectrum of data science applications.</p>

In [109]:
# Seed allows us to reporoduce the random numbers over and over again. 
#So, once run, the randomizer provides the same numbers everytime
np.random.seed(42)

# A single call to the uniform function that allows us to generate a float between 0 and 1
print(np.random.uniform())

# We can also pass in some parameters:
print(np.random.uniform(0,1e06))

#We can generate more than one number!
print(np.random.uniform(1,10,3))

print(np.random.randint(1,10, (6,6)))

0.3745401188473625
950714.3064099161
[7.58794548 6.38792636 2.40416776]
[[3 7 8 5 4 8]
 [8 3 6 5 2 8]
 [6 2 5 1 6 9]
 [1 3 7 4 9 3]
 [5 3 7 5 9 7]
 [2 4 9 2 9 5]]


## Working With CSV's

#### Imports

In [10]:
import csv

#### Reading a CSV

In [15]:
def open_csv(filepath, d = ','):
    #define a list to store the datain:
    data = []

    # use the with open and utf-8-sig encoding to load dat from .csv!
    with open(filepath, encoding='utf-8-sig') as my_data:
        info = csv.reader(my_data, delimiter=d)
        for row in info:
            data.append(row)
        return data

csv_data = open_csv(r'C:\Users\bioni\OneDrive\Documents\GitHub\Coding_Temple\Week 3\Data\redsox_2017_hitting.txt')
csv_data

[['Rk',
  'Pos',
  'Name',
  'Age',
  'G',
  'PA',
  'AB',
  'R',
  'H',
  '2B',
  '3B',
  'HR',
  'RBI',
  'SB',
  'CS',
  'BB',
  'SO',
  'BA',
  'OBP',
  'SLG',
  'OPS',
  'OPS+',
  'TB',
  'GDP',
  'HBP',
  'SH',
  'SF',
  'IBB'],
 ['1',
  'C',
  'Christian Vazquez',
  '26',
  '99',
  '345',
  '324',
  '43',
  '94',
  '18',
  '2',
  '5',
  '32',
  '7',
  '2',
  '17',
  '64',
  '.290',
  '.330',
  '.404',
  '.735',
  '91',
  '131',
  '14',
  '3',
  '0',
  '1',
  '0'],
 ['2',
  '1B',
  'Mitch Moreland',
  '31',
  '149',
  '576',
  '508',
  '73',
  '125',
  '34',
  '0',
  '22',
  '79',
  '0',
  '1',
  '57',
  '120',
  '.246',
  '.326',
  '.443',
  '.769',
  '99',
  '225',
  '14',
  '6',
  '0',
  '5',
  '6'],
 ['3',
  '2B',
  'Dustin Pedroia',
  '33',
  '105',
  '463',
  '406',
  '46',
  '119',
  '19',
  '0',
  '7',
  '62',
  '4',
  '3',
  '49',
  '48',
  '.293',
  '.369',
  '.392',
  '.760',
  '100',
  '159',
  '11',
  '2',
  '2',
  '4',
  '4'],
 ['4',
  'SS',
  'Xander Bogaerts',
  '

#### Loading a CSV's Data 

In [25]:
FIELDS = ['Rk','Pos','Name','Age','G','PA','AB','R','H','2B','3B','HR','RBI','SB',
          'CS','BB','SO','BA','OBP','SLG','OPS','OPS+','TB','GDP','HBP','SH','SF','IBB']
DATATYPES = [('rk', 'i'), ('pos', '|S25'), ('name', '|S25'), ('age', 'i'), ('g', 'i'), ('pa', 'i'), ('ab', 'i'),
                ('r', 'i'), ('h', 'i'), ('2b', 'i'), ('3b', 'i'), ('hr', 'i'), ('rbi', 'i'), ('sb', 'i'), ('cs', 'i'),
                ('bb', 'i'), ('so', 'i'), ('ba', 'f'), ('obp', 'f'), ('slg', 'f'), ('ops', 'f'), ('opsp', 'i'),
                ('tb', 'i'), ('gdp', 'i'), ('hbp', 'i'), ('sh', 'i'), ('sf', 'i'), ('ibb', 'i')]

def load_data(filepath, d = ','):
    data = np.genfromtxt(filepath,delimiter=d, skip_header=1, usecols = np.arange(0,24), invalid_raise=False, names=FIELDS, dtype=DATATYPES)
    return data
bs2017 = load_data(r'C:\Users\bioni\OneDrive\Documents\GitHub\Coding_Temple\Week 3\Data\redsox_2017_hitting.txt')

length=len(bs2017['BA'])

print(f'the average BA column is: {round((sum(bs2017["BA"])/length)*100,2)}')

#Total number of hits
print(f'Total Hits: {int(sum(bs2017["H"]))}')

#
print(f'Total Hits: {int(sum(bs2017["HR"]))/ int(sum(bs2017["H"]))}')

the average BA column is: 27.84
Total Hits: 1459
Total Hits: 0.11514736120630568


#### Summing the top 5 hitters for HR's

In [1]:
all_hr = bs2017['HR']

all_names = bs2017['Name']

new_names = []

for name in range(len(all_names)):
    new_names.append(all_names[name].decode('utf-8'))
zipped_values = list(zip(new_names, all_hr))

sorted_hr = sorted(zipped_values, key = lambda x: x[1], reverse= True)
sorted_hr[:5]

NameError: name 'bs2017' is not defined